File: src/remedy/gateway/channels/jwt_rs256.py
"""Minimal RS256 JWT verify + JWKS cache (stdlib only).

Used by Teams Bot Framework inbound auth. Not on the local chat/computer hot path.
JWKS is cached aggressively so webhook verification is one modular exponentiation
after the first fetch (~1–2ms), not a network round-trip every message.
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import logging
import threading
import time
import urllib.error
import urllib.request
from typing import Any

logger = logging.getLogger(__name__)

# SHA-256 DigestInfo prefix (PKCS#1 v1.5)
_DIGESTINFO_SHA256 = bytes.fromhex("3031300d060960864801650304020105000420")

# Bot Framework / Azure AD OpenID metadata (JWKS discovery).
_DEFAULT_OPENID_URLS = (
    "https://login.botframework.com/v1/.well-known/openidconfiguration",
    "https://login.microsoftonline.com/botframework.com/v2.0/.well-known/openid-configuration",
)

# Module-level JWKS cache: kid → (n_int, e_int)
_JWKS_LOCK = threading.Lock()
_JWKS: dict[str, tuple[int, int]] = {}
_JWKS_FETCHED_AT: float = 0.0
_JWKS_TTL_S = 6 * 3600  # 6h — BF keys rotate rarely; fail-open refresh on kid miss
# Test / offline inject: map kid → (n, e) without network
_TEST_KEYS: dict[str, tuple[int, int]] = {}


def b64url_decode(data: str) -> bytes:
    s = (data or "").encode("ascii") if isinstance(data, str) else data
    pad = b"=" * (-len(s) % 4)
    return base64.urlsafe_b64decode(s + pad)


def b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def clear_jwks_cache() -> None:
    """Test helper: drop cached network keys (keeps test injects)."""
    global _JWKS_FETCHED_AT
    with _JWKS_LOCK:
        _JWKS.clear()
        _JWKS_FETCHED_AT = 0.0


def inject_test_rsa_key(*, kid: str, n_b64: str, e_b64: str = "AQAB") -> None:
    """Register a public key for unit tests (no network)."""
    n = int.from_bytes(b64url_decode(n_b64), "big")
    e = int.from_bytes(b64url_decode(e_b64), "big")
    with _JWKS_LOCK:
        _TEST_KEYS[str(kid)] = (n, e)


def clear_test_rsa_keys() -> None:
    with _JWKS_LOCK:
        _TEST_KEYS.clear()


def _rsa_public_op(sig: bytes, n: int, e: int) -> bytes:
    k = (n.bit_length() + 7) // 8
    if len(sig) != k:
        # Allow missing leading zeros
        if len(sig) > k:
            return b""
        sig = sig.rjust(k, b"\x00")
    m = pow(int.from_bytes(sig, "big"), e, n)
    return m.to_bytes(k, "big")


def verify_rs256(token: str, *, n: int, e: int) -> bool:
    """Verify RS256 JWT signature with RSA public key (n, e)."""
    parts = (token or "").split(".")
    if len(parts) != 3:
        return False
    header_b64, payload_b64, sig_b64 = parts
    try:
        sig = b64url_decode(sig_b64)
    except Exception:
        return False
    signing_input = f"{header_b64}.{payload_b64}".encode("ascii")
    digest = hashlib.sha256(signing_input).digest()
    em = _rsa_public_op(sig, n, e)
    if not em:
        return False
    # PKCS#1 v1.5: 0x00 0x01 PS 0x00 DigestInfo
    if len(em) < 11 or em[0:2] != b"\x00\x01":
        return False
    try:
        sep = em.index(b"\x00", 2)
    except ValueError:
        return False
    ps = em[2:sep]
    if len(ps) < 8 or any(b != 0xFF for b in ps):
        return False
    digest_info = em[sep + 1 :]
    expected = _DIGESTINFO_SHA256 + digest
    # Exact match, constant time. Requiring the DigestInfo to be *exactly* the
    # tail is what rules out the classic e=3 forgery, where trailing garbage is
    # ignored; comparing in constant time is the standard on top of that.
    return hmac.compare_digest(digest_info, expected)


def decode_jwt_header(token: str) -> dict[str, Any] | None:
    parts = (token or "").split(".")
    if len(parts) != 3:
        return None
    try:
        data = json.loads(b64url_decode(parts[0]).decode("utf-8"))
        return data if isinstance(data, dict) else None
    except Exception:
        return None


def decode_jwt_payload_unverified(token: str) -> dict[str, Any] | None:
    parts = (token or "").split(".")
    if len(parts) != 3:
        return None
    try:
        data = json.loads(b64url_decode(parts[1]).decode("utf-8"))
        return data if isinstance(data, dict) else None
    except Exception:
        return None


_JWKS_HOSTS = frozenset(
    {
        "login.botframework.com",
        "login.microsoftonline.com",
        "login.microsoft.com",
        "login.windows.net",
    }
)


def _jwks_url_allowed(url: str) -> bool:
    from urllib.parse import urlparse

    parsed = urlparse((url or "").strip())
    if parsed.scheme != "https":
        return False
    host = (parsed.hostname or "").lower()
    if not host:
        return False
    if host in _JWKS_HOSTS:
        return True
    return host.endswith(".microsoftonline.com") or host.endswith(".windows.net")


def _http_get_json(url: str, *, timeout: float = 4.0) -> dict[str, Any] | None:
    if not _jwks_url_allowed(url):
        logger.warning("JWKS URL refused (host not allowlisted): %s", (url or "")[:80])
        return None
    try:
        from remedy.core.security import urlopen_no_redirect

        req = urllib.request.Request(
            url,
            headers={"Accept": "application/json", "User-Agent": "RemedyAI-TeamsJWT/1"},
            method="GET",
        )
        with urlopen_no_redirect(req, timeout=timeout) as resp:
            raw = resp.read(2 * 1024 * 1024)
        data = json.loads(raw.decode("utf-8"))
        return data if isinstance(data, dict) else None
    except (urllib.error.URLError, TimeoutError, json.JSONDecodeError, OSError) as exc:
        logger.warning("JWKS fetch failed %s: %s", url[:80], exc)
        return None


def _ingest_jwk_set(doc: dict[str, Any]) -> int:
    keys = doc.get("keys")
    if not isinstance(keys, list):
        return 0
    n_added = 0
    for jwk in keys:
        if not isinstance(jwk, dict):
            continue
        if str(jwk.get("kty") or "").upper() != "RSA":
            continue
        kid = str(jwk.get("kid") or "").strip()
        n_b64 = str(jwk.get("n") or "").strip()
        e_b64 = str(jwk.get("e") or "AQAB").strip()
        if not kid or not n_b64:
            continue
        try:
            n = int.from_bytes(b64url_decode(n_b64), "big")
            e = int.from_bytes(b64url_decode(e_b64), "big")
        except Exception:
            continue
        _JWKS[kid] = (n, e)
        n_added += 1
    return n_added


def refresh_jwks(*, force: bool = False, openid_urls: tuple[str, ...] | None = None) -> bool:
    """Fetch Bot Framework / Azure AD JWKS into the module cache.

    Returns True if at least one RSA key was loaded.
    """
    global _JWKS_FETCHED_AT
    now = time.time()
    with _JWKS_LOCK:
        if (
            not force
            and _JWKS
            and (now - _JWKS_FETCHED_AT) < _JWKS_TTL_S
        ):
            return True

    urls = openid_urls or _DEFAULT_OPENID_URLS
    jwks_uris: list[str] = []
    for meta_url in urls:
        meta = _http_get_json(meta_url)
        if not meta:
            continue
        juri = str(meta.get("jwks_uri") or "").strip()
        if _jwks_url_allowed(juri):
            jwks_uris.append(juri)
    # Also try well-known keys endpoints directly (fast path when meta is slow)
    jwks_uris.extend(
        [
            "https://login.botframework.com/v1/.well-known/keys",
            "https://login.microsoftonline.com/common/discovery/v2.0/keys",
        ]
    )
    seen: set[str] = set()
    total = 0
    with _JWKS_LOCK:
        for uri in jwks_uris:
            if uri in seen:
                continue
            seen.add(uri)
            doc = _http_get_json(uri)
            if doc:
                total += _ingest_jwk_set(doc)
        if total:
            _JWKS_FETCHED_AT = time.time()
            logger.info("Teams JWKS loaded: %d RSA keys", len(_JWKS))
            return True
    return bool(_JWKS)


def _lookup_key(kid: str) -> tuple[int, int] | None:
    with _JWKS_LOCK:
        if kid in _TEST_KEYS:
            return _TEST_KEYS[kid]
        if kid in _JWKS:
            return _JWKS[kid]
    return None


def verify_jwt_rs256_jwks(
    token: str,
    *,
    allow_network: bool = True,
) -> bool:
    """Verify JWT with RS256 using cached (or freshly fetched) JWKS.

    Network is only used when the kid is unknown or the cache is empty/expired.
    """
    header = decode_jwt_header(token)
    if not header:
        return False
    alg = str(header.get("alg") or "").upper()
    if alg != "RS256":
        # Bot Framework uses RS256; reject others fail-closed
        logger.warning("Teams JWT alg not RS256: %s", alg)
        return False
    kid = str(header.get("kid") or "").strip()
    if not kid:
        logger.warning("Teams JWT missing kid")
        return False

    key = _lookup_key(kid)
    if key is None and allow_network:
        refresh_jwks(force=False)
        key = _lookup_key(kid)
        if key is None:
            # kid miss after soft refresh — force once
            refresh_jwks(force=True)
            key = _lookup_key(kid)
    if key is None:
        logger.warning("Teams JWT kid not in JWKS: %s", kid[:40])
        return False
    n, e = key
    if not verify_rs256(token, n=n, e=e):
        logger.warning("Teams JWT RS256 signature invalid")
        return False
    return True

File: native/go/gateway/jwt_rs256.go
package gateway

import (
	"crypto"
	"crypto/rsa"
	"crypto/sha256"
	"encoding/base64"
	"encoding/json"
	"io"
	"log"
	"math/big"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// jwksSource is one issuer family with its own key cache. Keys from one
// source are never used to verify tokens attributed to another.
type jwksSource struct {
	name string
	// openID metadata documents whose jwks_uri is followed (optional).
	openID []string
	// direct JWKS documents fetched after metadata discovery.
	direct []string
	// hosts allowed for any JWKS / metadata fetch of this source.
	hosts       map[string]struct{}
	hostSuffix  []string
	mu          sync.Mutex
	keys        map[string]*rsa.PublicKey
	fetchedAt   time.Time
	logPrefix   string
	verifyLabel string
}

var (
	// teamsJWKS covers Bot Framework / Azure AD issued tokens.
	teamsJWKS = &jwksSource{
		name: "teams",
		openID: []string{
			"https://login.botframework.com/v1/.well-known/openidconfiguration",
			"https://login.microsoftonline.com/botframework.com/v2.0/.well-known/openid-configuration",
		},
		direct: []string{
			"https://login.botframework.com/v1/.well-known/keys",
			"https://login.microsoftonline.com/common/discovery/v2.0/keys",
		},
		hosts: map[string]struct{}{
			"login.botframework.com":    {},
			"login.microsoftonline.com": {},
			"login.microsoft.com":       {},
			"login.windows.net":         {},
		},
		hostSuffix: []string{".microsoftonline.com", ".windows.net"},
		keys:       map[string]*rsa.PublicKey{},
		logPrefix:  "teams",
	}

	// googleChatJWKS covers tokens minted by the Google Chat service account.
	googleChatJWKS = &jwksSource{
		name:      "google_chat",
		direct:    []string{googleChatJWKSURL},
		hosts:     map[string]struct{}{"www.googleapis.com": {}},
		keys:      map[string]*rsa.PublicKey{},
		logPrefix: "google_chat",
	}

	jwksTTL  = 6 * time.Hour
	jwksHTTP = &http.Client{Timeout: 4 * time.Second}

	jwksTestMu   sync.Mutex
	jwksTestKeys = map[string]*rsa.PublicKey{}
)

// GoogleChatIssuer is the service account Google Chat signs webhook JWTs with.
const GoogleChatIssuer = "chat@system.gserviceaccount.com"

const googleChatJWKSURL = "https://www.googleapis.com/service_accounts/v1/jwk/" + GoogleChatIssuer

// ClearJWKSCache drops cached network keys (keeps test injects).
func ClearJWKSCache() {
	for _, src := range []*jwksSource{teamsJWKS, googleChatJWKS} {
		src.mu.Lock()
		src.keys = map[string]*rsa.PublicKey{}
		src.fetchedAt = time.Time{}
		src.mu.Unlock()
	}
}

// InjectTestRSAKey registers a public key for unit tests (no network).
func InjectTestRSAKey(kid, nB64, eB64 string) {
	if eB64 == "" {
		eB64 = "AQAB"
	}
	nBytes, err := b64URLDecode(nB64)
	if err != nil {
		return
	}
	eBytes, err := b64URLDecode(eB64)
	if err != nil {
		return
	}
	pub := &rsa.PublicKey{
		N: new(big.Int).SetBytes(nBytes),
		E: int(new(big.Int).SetBytes(eBytes).Int64()),
	}
	jwksTestMu.Lock()
	jwksTestKeys[kid] = pub
	jwksTestMu.Unlock()
}

// ClearTestRSAKeys removes injected test keys.
func ClearTestRSAKeys() {
	jwksTestMu.Lock()
	defer jwksTestMu.Unlock()
	jwksTestKeys = map[string]*rsa.PublicKey{}
}

func b64URLDecode(data string) ([]byte, error) {
	s := data
	switch len(s) % 4 {
	case 2:
		s += "=="
	case 3:
		s += "="
	}
	return base64.URLEncoding.DecodeString(s)
}

// DecodeJWTPayloadUnverified decodes the JWT payload without signature verify.
func DecodeJWTPayloadUnverified(token string) map[string]any {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil
	}
	raw, err := b64URLDecode(parts[1])
	if err != nil {
		return nil
	}
	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil
	}
	return data
}

func decodeJWTHeader(token string) map[string]any {
	parts := strings.Split(token, ".")
	if len(parts) != 3 {
		return nil
	}
	raw, err := b64URLDecode(parts[0])
	if err != nil {
		return nil
	}
	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil
	}
	return data
}

func (src *jwksSource) urlAllowed(raw string) bool {
	u, err := url.Parse(strings.TrimSpace(raw))
	if err != nil || !strings.EqualFold(u.Scheme, "https") {
		return false
	}
	host := strings.ToLower(u.Hostname())
	if host == "" {
		return false
	}
	if _, ok := src.hosts[host]; ok {
		return true
	}
	for _, suf := range src.hostSuffix {
		if strings.HasSuffix(host, suf) {
			return true
		}
	}
	return false
}

func (src *jwksSource) getJSON(rawURL string) map[string]any {
	if !src.urlAllowed(rawURL) {
		log.Printf("%s: JWKS URL refused (host not allowlisted): %s", src.logPrefix, trimRunes(rawURL, 80))
		return nil
	}
	req, err := http.NewRequest(http.MethodGet, rawURL, nil)
	if err != nil {
		return nil
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "RemedyAI-JWT/1")
	resp, err := jwksHTTP.Do(req)
	if err != nil {
		log.Printf("%s: JWKS fetch failed %s: %v", src.logPrefix, trimRunes(rawURL, 80), err)
		return nil
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(io.LimitReader(resp.Body, 2<<20))
	if err != nil {
		return nil
	}
	var data map[string]any
	if err := json.Unmarshal(raw, &data); err != nil {
		return nil
	}
	return data
}

// ingestJWKSet parses RSA keys from a JWK set into dst.
func ingestJWKSet(dst map[string]*rsa.PublicKey, doc map[string]any) int {
	keys, _ := doc["keys"].([]any)
	nAdded := 0
	for _, raw := range keys {
		jwk, _ := raw.(map[string]any)
		if jwk == nil {
			continue
		}
		if !strings.EqualFold(anyString(jwk["kty"]), "RSA") {
			continue
		}
		kid := strings.TrimSpace(anyString(jwk["kid"]))
		nB64 := strings.TrimSpace(anyString(jwk["n"]))
		eB64 := strings.TrimSpace(anyString(jwk["e"]))
		if eB64 == "" {
			eB64 = "AQAB"
		}
		if kid == "" || nB64 == "" {
			continue
		}
		nBytes, err := b64URLDecode(nB64)
		if err != nil {
			continue
		}
		eBytes, err := b64URLDecode(eB64)
		if err != nil {
			continue
		}
		dst[kid] = &rsa.PublicKey{
			N: new(big.Int).SetBytes(nBytes),
			E: int(new(big.Int).SetBytes(eBytes).Int64()),
		}
		nAdded++
	}
	return nAdded
}

// refresh fetches the JWKS documents for this source into its cache.
func (src *jwksSource) refresh(force bool) bool {
	now := time.Now()
	src.mu.Lock()
	if !force && len(src.keys) > 0 && now.Sub(src.fetchedAt) < jwksTTL {
		src.mu.Unlock()
		return true
	}
	src.mu.Unlock()

	uris := make([]string, 0, 8)
	for _, metaURL := range src.openID {
		meta := src.getJSON(metaURL)
		if meta == nil {
			continue
		}
		juri := strings.TrimSpace(anyString(meta["jwks_uri"]))
		if src.urlAllowed(juri) {
			uris = append(uris, juri)
		}
	}
	uris = append(uris, src.direct...)
	seen := map[string]struct{}{}
	docs := make([]map[string]any, 0, len(uris))
	for _, uri := range uris {
		if _, ok := seen[uri]; ok {
			continue
		}
		seen[uri] = struct{}{}
		if doc := src.getJSON(uri); doc != nil {
			docs = append(docs, doc)
		}
	}
	src.mu.Lock()
	defer src.mu.Unlock()
	total := 0
	for _, doc := range docs {
		total += ingestJWKSet(src.keys, doc)
	}
	if total > 0 {
		src.fetchedAt = time.Now()
		log.Printf("%s: JWKS loaded: %d RSA keys", src.logPrefix, len(src.keys))
		return true
	}
	return len(src.keys) > 0
}

func (src *jwksSource) lookup(kid string) *rsa.PublicKey {
	jwksTestMu.Lock()
	if k, ok := jwksTestKeys[kid]; ok {
		jwksTestMu.Unlock()
		return k
	}
	jwksTestMu.Unlock()
	src.mu.Lock()
	defer src.mu.Unlock()
	return src.keys[kid]
}

// verifyToken checks alg/kid, resolves the key (network when allowed) and
// verifies the RS256 signature.
func (src *jwksSource) verifyToken(token string, allowNetwork bool) bool {
	header := decodeJWTHeader(token)
	if header == nil {
		return false
	}
	alg := strings.ToUpper(anyString(header["alg"]))
	if alg != "RS256" {
		log.Printf("%s: JWT alg not RS256: %s", src.logPrefix, alg)
		return false
	}
	kid := strings.TrimSpace(anyString(header["kid"]))
	if kid == "" {
		log.Printf("%s: JWT missing kid", src.logPrefix)
		return false
	}
	key := src.lookup(kid)
	if key == nil && allowNetwork {
		src.refresh(false)
		key = src.lookup(kid)
		if key == nil {
			src.refresh(true)
			key = src.lookup(kid)
		}
	}
	if key == nil {
		log.Printf("%s: JWT kid not in JWKS: %s", src.logPrefix, trimRunes(kid, 40))
		return false
	}
	if !verifyRS256(token, key) {
		log.Printf("%s: JWT RS256 signature invalid", src.logPrefix)
		return false
	}
	return true
}

// RefreshJWKS fetches Bot Framework / Azure AD JWKS into the cache.
func RefreshJWKS(force bool) bool {
	return teamsJWKS.refresh(force)
}

func verifyRS256(token string, pub *rsa.PublicKey) bool {
	parts := strings.Split(token, ".")
	if len(parts) != 3 || pub == nil {
		return false
	}
	sig, err := b64URLDecode(parts[2])
	if err != nil {
		return false
	}
	signingInput := parts[0] + "." + parts[1]
	sum := sha256.Sum256([]byte(signingInput))
	return rsa.VerifyPKCS1v15(pub, crypto.SHA256, sum[:], sig) == nil
}

// VerifyJWTRS256JWKS verifies a Bot Framework JWT with RS256 against the
// Teams JWKS cache (network fetch when allowNetwork).
func VerifyJWTRS256JWKS(token string, allowNetwork bool) bool {
	return teamsJWKS.verifyToken(token, allowNetwork)
}

// VerifyGoogleChatJWT verifies a Google Chat webhook bearer token:
// RS256 signature against the chat service account JWKS, iss equal to that
// account, aud equal to the configured Cloud project number, and exp in the
// future (60 s skew). Fail closed on any missing claim.
func VerifyGoogleChatJWT(token, projectNumber string, now time.Time, allowNetwork bool) bool {
	projectNumber = strings.TrimSpace(projectNumber)
	if projectNumber == "" || strings.TrimSpace(token) == "" {
		return false
	}
	claims := DecodeJWTPayloadUnverified(token)
	if claims == nil {
		return false
	}
	if iss := strings.TrimSpace(anyString(claims["iss"])); iss != GoogleChatIssuer {
		log.Printf("google_chat: JWT iss rejected: %s", trimRunes(iss, 80))
		return false
	}
	audOK := false
	switch v := claims["aud"].(type) {
	case string:
		audOK = strings.TrimSpace(v) == projectNumber
	case []any:
		for _, a := range v {
			if anyString(a) == projectNumber {
				audOK = true
				break
			}
		}
	}
	if !audOK {
		log.Printf("google_chat: JWT aud does not match project_number")
		return false
	}
	if now.IsZero() {
		now = time.Now()
	}
	exp, ok := anyFloat(claims["exp"])
	if !ok || float64(now.Unix()) >= exp+60 {
		log.Printf("google_chat: JWT expired or missing exp")
		return false
	}
	if nbf, ok := anyFloat(claims["nbf"]); ok && float64(now.Unix())+60 < nbf {
		return false
	}
	return googleChatJWKS.verifyToken(token, allowNetwork)
}
